Sources/WASI/FileSystem.swift

```diff
@@ -120,3 +127,55 @@ struct FdTable {
         }
     }
 }
+
+/// Content of a file that can be retrieved from the file system.
+public enum FileContent {
+    case bytes([UInt8])
+    case handle(FileDescriptor)
+}
+
+/// Public protocol for file system providers that users interact with.
+///
+/// This protocol exposes only user-facing methods for managing files and directories.
+public protocol FileSystemProvider {
+    /// Adds a file to the file system with the given byte content.
+    func addFile(at path: String, content: [UInt8]) throws
+
+    /// Adds a file to the file system with the given string content.
+    func addFile(at path: String, content: String) throws
+
+    /// Adds a file to the file system backed by a file descriptor handle.
+    func addFile(at path: String, handle: FileDescriptor) throws
+
+    /// Gets the content of a file at the specified path.
+    func getFile(at path: String) throws -> FileContent
+
+    /// Removes a file from the file system.
+    func removeFile(at path: String) throws
+}
+
+/// Internal protocol for file system implementations used by WASI.
+///
+/// This protocol contains WASI-specific implementation details that should not
+/// be exposed to library users.
+internal protocol FileSystem {
+    /// Returns the list of pre-opened directory paths.
+    func getPreopenPaths() -> [String]
```

Comment:
`get`-prefixed functions are not necessarily idiomatic in Swift, I'd suggest converting this to a ready-only property. 

```suggestion
    var preopenPaths: [String] { get }
```